Packer Builder ProfitBricks

jasminSPC opened the pull request:
This pull request contains code for [ProfitBricks](https://profitbricks.com/) builder.

jasminSPC commented:
@rickard-von-essen I see that you labeled this PR "need-documentation" I've already have documentation but I'm not sure where to put it.

rickard-von-essen commented:
The docs for builders goes into https://github.com/mitchellh/packer/tree/master/website/source/docs/builders

rickard-von-essen commented:
Also there is some commented out debug code and some copy paste comments that are inaccurate. Please clean such things

jasminSPC commented:
@rickard-von-essen I've addressed your comments. 
What is the next step?

rickard-von-essen commented:
Someone needs to do some basic verification. And Hashicorp has to decide if they think this should be in packer core. 

I have registered a ProfitBricks account, but not yeet got access and currently I'm on vacation and don't have much time at the keyboard, so if no one else picks it up there will be some weeks (at least) before I can provide any feedback. 

@cbednarski any quick comments?

jasminSPC commented:
@rickard-von-essen Thanks. If by Monday you don't get ProfitBricks credentials please let us know.

cbednarski commented:
@jasminSPC Sorry to keep you waiting on a reply to this.

We prioritize core features based on community demand and the availability of resources to maintain ongoing support. For example, we recently added support for Azure to Packer and Terraform and have two contributors from Microsoft and at least one community expert who are helping us keep pace with customer requests and the evolving upstream APIs.

I recommend that you reach out via our [partner channel](https://www.hashicorp.com/partners.html) (see the end of the page for contact info) so we can kick off this sort of conversation.

In the short term the best delivery option is for you to ship this as an third-party plugin. Since you have built this in the packer codebase there is minimal work for you to do this. See [plugin/example/main.go](https://github.com/mitchellh/packer/blob/master/plugin/example/main.go) for details on how to build a plugin binary.

To deliver this to a customer you can simply provide them instructions to install it in the same directory where packer is installed. Up until Packer 0.8.6 all official plugins were distributed this way and it was only changed recently to save space -- otherwise the functionality is identical. If you need more examples you can check back in git history at version 0.8.6 and earlier.

I am going to close this for now but please reach out via our partner channel so we can follow up. Thanks!

cbednarski commented:
@jasminSPC I resolved the merge conflict and stubbed out the documentation in a [branch](https://github.com/mitchellh/packer/tree/f-profitbricks). Can you incorporate these into your PR, and fill out the rest of the docs?

cbednarski commented:
Rename `builder/profitbricks/artefact.go` -> `builder/profitbricks/artifact.go`

cbednarski commented on builder/profitbricks/builder_acc_test.go:
```diff
@@ -0,0 +1,38 @@
+	      "image": "Ubuntu-16.04",
+	      "pbpassword": "password",
+	      "pbusername": "username",
+	      "servername": "packer",
```
The documentation / examples should reflect these keys (`pbpassword`, `pbusername`, and `servername`).

mwhooker replied:
same as above

cbednarski commented on builder/profitbricks/builder_test.go:
```diff
@@ -0,0 +1,57 @@
+	return map[string]interface{}{
+		"image": "Ubuntu-16.04",
+		"password": "password",
+		"username": "username",
```
`testBuilderAccBasic` uses `pbusername` and `pbpassword` but this test uses `username` and `password`. Which is correct? Are these different credentials? If so please add them to the docs.

cbednarski commented on vendor/github.com/profitbricks/profitbricks-sdk-go/req.go:
```diff
@@ -0,0 +1,128 @@
+	if strings.HasPrefix(path, "http") {
+		//REMOVE AFTER TESTING
+		path := strings.Replace(path, "https://api.profitbricks.com/rest/v2", Endpoint, 1)
+		// END REMOVE
```
Should this still be here?

jasminSPC replied:
This is just a comment in the SDK. The if above will replace "http" path with a default REST path.

cbednarski commented on vendor/github.com/profitbricks/profitbricks-sdk-go/req.go:
```diff
@@ -0,0 +1,128 @@
+	}
+	if strings.HasPrefix(path, "<base>") {
+		//REMOVE AFTER TESTING
+		path := strings.Replace(path, "<base>", Endpoint, 1)
```
Should this still be here? If you need a value set for tests only, use an environment variable.

jasminSPC replied:
Same as above

cbednarski commented on builder/profitbricks/config.go:
```diff
@@ -0,0 +1,129 @@
+	}
+
+	if c.PBUsername == "" {
+		c.PBUsername = os.Getenv("PROFITBRICKS_USERNAME")
```
Please document that these options can be specified by environment instead, and indicate that if both are specified the value in the file will be used.

jasminSPC replied:
I've updated the docs here:
https://github.com/mitchellh/packer/pull/3660/files#diff-15d79ed77362bc4d0fb13ddebcf1a2f0R28
and here
https://github.com/mitchellh/packer/pull/3660/files#diff-15d79ed77362bc4d0fb13ddebcf1a2f0R30

jasminSPC commented:
I'm having problems withi this:

```
command/plugin.go:43:2: cannot find package "github.com/mitchellh/packer/post-processor/manifest" in any of:
        /Users/jasmin/go/src/github.com/StackPointCloud/packer/vendor/github.com/mitchellh/packer/post-processor/manifest (vendor tree)
        /usr/local/Cellar/go/1.6/libexec/src/github.com/mitchellh/packer/post-processor/manifest (from $GOROOT)
        /Users/jasmin/go/src/github.com/mitchellh/packer/post-processor/manifest (from $GOPATH)
```

Can you help me out?

jasminSPC commented:
for some reason **github.com/approvals/go-approval-tests** that are used by **/packer/builder/azure/common/template/template_builder_test.go** was not added to the vendor folder so I had to add them manually.

jasminSPC commented:
I've made quite few commits in this PR. Should I squash them before we proceed?

jasminSPC commented:
@rickard-von-essen Are there any other remarks on this PR?

rickard-von-essen commented on website/source/docs/builders/profitbricks.html.md:
```diff
@@ -0,0 +1,75 @@
+
+The ProfitBricks Builder is able to create virtual machines for [ProfitBricks](https://www.profitbricks.com).
+
+-> **Note:** This builder is not supported by HashiCorp. Please visit [ProfitBricks DevOps Central](https://devops.profitbricks.com/) for support. You may file issues on [GitHub](https://github.com/profitbricks/docker-machine-driver-profitbricks/issues).
```
Is the url intended to be https://github.com/profitbricks/docker-machine-driver-profitbricks/issues ?

rickard-von-essen commented on website/source/docs/builders/profitbricks.html.md:
```diff
@@ -0,0 +1,75 @@
+
+### Optional
+
+-   `cores` (int) - Amount of CPU cores to use for this build. Defaults to `4`.
```
int -> integer, in several places.
Default values should have be double quoted `4` -> "4", in several places.

rickard-von-essen commented on website/source/docs/builders/profitbricks.html.md:
```diff
@@ -0,0 +1,75 @@
+
+-   `snapshot_name` (string) - If snapshot name is not provided Packer will generate it
+
+-   `snapshot_password` (string) - ????
```
TODO

rickard-von-essen commented on website/source/docs/builders/profitbricks.html.md:
```diff
@@ -0,0 +1,75 @@
+
+-   `snapshot_password` (string) - ????
+
+-   `ssh_key_path` (string) - Path to private SSHkey. If no path to the key is provided Packer will create one under the name [snapshot_name]
```
This seems to duplicate https://www.packer.io/docs/templates/communicator.html#ssh_private_key_file . If it actually is something different I think it would be good with a less confusing name and some more details in the docs.

References to other configuration options should be written `snapshot_name` instead of [snapshot_name].

rickard-von-essen commented on website/source/docs/builders/profitbricks.html.md:
```diff
@@ -0,0 +1,75 @@
+
+-   `ssh_key_path` (string) - Path to private SSHkey. If no path to the key is provided Packer will create one under the name [snapshot_name]
+
+-   `url` (string) - Endpoint for the ProfitBricks REST API
```
Maybe something about the default.

rickard-von-essen commented on website/source/docs/builders/profitbricks.html.md:
```diff
@@ -0,0 +1,75 @@
+
+-   `url` (string) - Endpoint for the ProfitBricks REST API
+
+-   `timeout` (string) - An approximate limit on how long packer will continue making status requests while waiting for the build to complete. Set this to an integer where 1 equals 10 seconds. Default value 50.
```
Capital P in packer.

rickard-von-essen replied:
If 1 equals 1 s. what does 50 equals? This needs more explanation or a link to the ProfitBricks docs.

rickard-von-essen commented on website/source/docs/builders/profitbricks.html.md:
```diff
@@ -0,0 +1,75 @@
+
+There are many configuration options available for the builder. They are
+segmented below into two categories: required and optional parameters. Within
+each category, the available configuration keys are alphabetized.
```
_Alphabetize_ - Fix order of options

rickard-von-essen commented:
@jasminSPC  could you drop me an email.

jasminSPC commented:
jasmin@stackpointcloud.com

jasminSPC commented:
I've addressed the comments above and I've synced the fork.

rickard-von-essen commented on website/source/docs/builders/profitbricks.html.md:
```diff
@@ -0,0 +1,74 @@
+
+-   `cores` (integer) - Amount of CPU cores to use for this build. Defaults to "4".
+
+-   `disk_size` (string) - Amount of disk space for this image. Defaults to "50gb"
```
When setting "10gb" it complains that it should be an integer.

rickard-von-essen commented:
Tried to use this but a got this [crash on first try](https://gist.github.com/anonymous/39b5e33f5da36ac8aef6089b62b466d4).

rickard-von-essen commented on website/source/docs/builders/profitbricks.html.md:
```diff
@@ -0,0 +1,74 @@
+
+The ProfitBricks Builder is able to create virtual machines for [ProfitBricks](https://www.profitbricks.com).
+
+-> **Note:** This builder is not supported by HashiCorp. Please visit [ProfitBricks DevOps Central](https://devops.profitbricks.com/) for support. You may file issues on [GitHub](https://github.com/profitbricks/packer-builder-profitbricks/issues).
```
@cbednarski Is this how we would like it?

rickard-von-essen replied:
@mwhooker I think this is not a good way to handle this. I think ProfitBricks and HashiCorp should sort out how to handle maintenance.

jasminSPC commented:
@rickard-von-essen

It is a limitation by the provider. ProfitBricks allows sshKeys on Linux images.

```
==> profitbricks:   "httpStatus" : 422,
==> profitbricks:   "messages" : [ {
==> profitbricks:     "errorCode" : "100",
==> profitbricks:     "message" : "[(root).entities.servers.items.[0].entities.volumes.items.[0].properties.image] Passwords/SSH Keys are only accepted for public ProfitBricks Images."
==> profitbricks:   }, {
==> profitbricks:     "errorCode" : "100",
==> profitbricks:     "message" : "[(root).entities.servers.items.[0].entities.volumes.items.[0].properties.sshKeys] SSH keys are only supported for ProfitBricks Linux images."
==> profitbricks:   } ]
==> profitbricks: }
==> profitbricks:[0m
```

I would suggest you try Ubuntu.

rickard-von-essen commented:
Ok, but the error should be handled gracefully. Where can I find the Ubuntu image?

jasminSPC commented:
The Ubuntu 16 image is the default image as seen [here](https://github.com/mitchellh/packer/pull/3660/files#diff-02446bdb5d260cc008f9bf8f729ac82aR84
). You can pass the UUID as well as the full or partial name of the image. For example, `debian-8`, `centos-7` or `ubuntu-14`.

There is no definitive list of images that support SSH key injection as far as I know. 
The only place where you can get the list of all available images is [here](https://api.profitbricks.com/rest/v2/images?depth=5) or you could use the [ProfitBricks CLI](https://devops.profitbricks.com/tools/cli/).

Lastly, the error 422 is a generic validation error code and it can mean number of things. I'm not quite sure how to handle unsupported SSH keys gracefully.

rickard-von-essen commented on website/source/docs/builders/profitbricks.html.md:
```diff
@@ -0,0 +1,74 @@
+
+### Required
+
+-   `image` (string) - ProfitBricks volume image 
```
This should state the default value.

rickard-von-essen commented:
> There is no definitive list of images that support SSH key injection as far as I know. 
The only place where you can get the list of all available images is here or you could use the ProfitBricks CLI.

Maybe add a note to the docs about how to find public images with the cli and some details about using ssh key injection vs not.

rickard-von-essen commented:
> Lastly, the error 422 is a generic validation error code and it can mean number of things. I'm not quite sure how to handle unsupported SSH keys gracefully.

So this part returning an error is fine (even if it would be nice if was parsed and returned a bit human friendly if possible*):

```
2016/09/12 16:02:13 ui error: [1;31m==> profitbricks: {
==> profitbricks:   "httpStatus" : 422,
==> profitbricks:   "messages" : [ {
==> profitbricks:     "errorCode" : "100",
==> profitbricks:     "message" : "[(root).entities.servers.items.[0].entities.volumes.items.[0].properties.image] Passwords/SSH Keys are only accepted for public ProfitBricks Images."
==> profitbricks:   }, {
==> profitbricks:     "errorCode" : "100",
==> profitbricks:     "message" : "[(root).entities.servers.items.[0].entities.volumes.items.[0].properties.sshKeys] SSH keys are only supported for ProfitBricks Linux images."
==> profitbricks:   } ]
==> profitbricks: }
==> profitbricks:[0m
2016/09/12 16:02:13 ui: [1;32m==> profitbricks: Removing Virtual Data Center...[0m
```

But panic if not ok. 
```
2016/09/12 16:02:13 packer: panic: interface conversion: interface is nil, not string
2016/09/12 16:02:13 packer: 
2016/09/12 16:02:13 packer: goroutine 51 [running]:
2016/09/12 16:02:13 packer: panic(0xbc1d80, 0xc4204c8640)
```

----
*) It would be nice if it read something like:
```
2016/09/12 16:02:13 ui error: [1;31m==> profitbricks: {
==> profitbricks:   http status [422]:
==> profitbricks:     error: 100 [(root).entities.servers.items.[0].entities.volumes.items.[0].properties.image] Passwords/SSH Keys are only accepted for public ProfitBricks Images.
==> profitbricks:     error: 100 [(root).entities.servers.items.[0].entities.volumes.items.[0].properties.sshKeys] SSH keys are only supported for ProfitBricks Linux images.
==> profitbricks:[0m
2016/09/12 16:02:13 ui: [1;32m==> profitbricks: Removing Virtual Data Center...[0m
```

rickard-von-essen commented:
Ok I successfully built a Ubuntu 16.04 image. It would be useful if there where a link in the docs to instructions on how ssh key injection works (is it just cloud-init?) etc.

jasminSPC commented:
The error messages will look like this:
```==> profitbricks: [(root).entities.servers.items.[0].entities.volumes.items.[0].properties.image] Image with uuid 31d779f6-75ee-11e6-bfbf-52540005ab80 has a size 13 GB which is greater than the volume size 5 GB.
==> profitbricks: [(root).entities.servers.items.[0].entities.volumes.items.[0].properties.sshKeys] SSH keys are only supported for ProfitBricks Linux images.
==> profitbricks:
```

jasminSPC commented:
@rickard-von-essen I've addressed all the remarks above. For image parameter i've put this
```
-   `image` (string) - ProfitBricks volume image. Only Linux public images are supported. Defaults to "Ubuntu-16.04". To obtain full list of available images you can use [ProfitBricks CLI](https://github.com/profitbricks/profitbricks-cli#image). 
```

jasminSPC commented:
@rickard-von-essen the CI is failing for this `FAIL  github.com/mitchellh/packer/builder/amazon/common	0.098s` I'm not sure if I should do anything there?

rickard-von-essen commented:
No you can ignore the test failures, it's my fault. I'll fix it tonight.

jasminSPC commented:
Alright
Let me know if you need anything else.

Jasmin Gacić

On Sep 14, 2016 15:55, "Rickard von Essen" <notifications@github.com> wrote:

> No you can ignore the test failures, it's my fault. I'll fix it tonight.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/mitchellh/packer/pull/3660#issuecomment-247020192>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AH8SR8enFPP6kbPfal4lgGqmbyQi3Tneks5qp_ymgaJpZM4JAt8T>
> .
>

jasminSPC commented:
Are there any updates?

rickard-von-essen commented:
ping @mwhooker

mwhooker commented:
HI there,

definitely want to get this in for the 0.11 release. It looks like @rickard-von-essen might have some changes to fix the tests? Let me know if I can help.

will review in the meantime

mwhooker commented:
@jasminSPC could you shoot me an email @ mhooker at hashicorp? want to sync up before merging this

jasminSPC commented:
Renamed artefact -> artifact as requested
An example is located here https://github.com/mitchellh/packer/pull/3660/files#diff-15d79ed77362bc4d0fb13ddebcf1a2f0R58
Fixed tests.

And I'm not sure about maintenance plane.

jasminSPC commented:
Change to profitbricks.html.md has been pushed.

jasminSPC commented:
@mwhooker I've submited all the changes you requested but it seems the stuff doesn't change under "Changes requested"

mwhooker commented on website/source/docs/builders/profitbricks.html.md:
```diff
@@ -0,0 +1,72 @@
+
+-   `image` (string) - ProfitBricks volume image. Only Linux public images are supported. Defaults to "Ubuntu-16.04". To obtain full list of available images you can use [ProfitBricks CLI](https://github.com/profitbricks/profitbricks-cli#image). 
+
+-   `password` (string) - ProfitBrucks password. This can be specified via environment variable `PROFITBRICKS_PASSWORD', if provided. The value definded in the config has precedence over environemnt variable.
```
s/profitbrucks/profitbricks/

jasminSPC replied:
Done

mwhooker commented on website/source/docs/builders/profitbricks.html.md:
```diff
@@ -0,0 +1,72 @@
+
+-   `cores` (integer) - Amount of CPU cores to use for this build. Defaults to "4".
+
+-   `disk_size` (string) - Amount of disk space for this image. Defaults to "50"
```
Can you specify the units here? I assume MB, but want to make it obvious

jasminSPC replied:
Done

mwhooker commented:
see also https://github.com/hashicorp/terraform/pull/7943

jasminSPC commented:
Now when the plugin crashes the out put will look like this:
 
==> profitbricks: Creating temporary SSH key for instance...
    profitbricks: Saving key to: pb_packer-1476311686
==> profitbricks: Creating Virtual Data Center...
**==> profitbricks: Unauthorized**
==> profitbricks: Removing Virtual Data Center...
Build 'profitbricks' finished.

jasminSPC commented:
@mwhooker When you say https://github.com/mitchellh/packer/pull/3660#issuecomment-253313827 what do you mean?

mwhooker commented:
@jasminSPC the terraform link? just putting it there for reference, since when we merge one we'll probably merge the other

jasminSPC commented:
@mwhooker As a reminder I guess :)

baldwinSPC commented:
@mwhooker pinging on this PR, too.

jasminSPC commented:
@mwhooker is there anything further required that we do?
